test/bench/synthetic_benchmarks.cpp

```diff
@@ -0,0 +1,202 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2020 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "synthetic_benchmarks.hpp"
+#include "helpers.hpp"
+#include "test/utils/bytecode.hpp"
+#include <evmc/instructions.h>
+#include <evmone/instruction_traits.hpp>
+
+using namespace benchmark;
+
+namespace evmone::test
+{
+namespace
+{
+/// Stack limit inside the EVM benchmarking loop (one stack item is used for the loop count).
+constexpr auto stack_limit = 1023;
+
+enum class Mode
+{
+    min_stack = 0,   ///< The code uses as minimal stack as possible.
+    full_stack = 1,  ///< The code fills the stack up to its limit.
+};
+
+enum class InstructionKind : char
+{
+    nullop = 'n',
+    unop = 'u',
+    binop = 'b',
+    push = 'p',
+    swap = 's',
+    dup = 'd',
+    producer = 'a',
+    unknown = 'X',
+};
+
+constexpr InstructionKind get_instruction_kind(evmc_opcode opcode) noexcept
+{
+    const auto trait = instr::traits[opcode];
+    if (opcode >= OP_PUSH1 && opcode <= OP_PUSH32)
+        return InstructionKind::push;
+    else if (opcode >= OP_SWAP1 && opcode <= OP_SWAP16)
+        return InstructionKind::swap;
+    else if (opcode >= OP_DUP1 && opcode <= OP_DUP16)
+        return InstructionKind::dup;
+    else if (trait.stack_height_required == 0 && trait.stack_height_change == 0)
+        return InstructionKind::nullop;
+    else if (trait.stack_height_required == 1 && trait.stack_height_change == 0)
+        return InstructionKind::unop;
+    else if (trait.stack_height_required == 2 && trait.stack_height_change == -1)
+        return InstructionKind::binop;
+    else if (trait.stack_height_required == 0 && trait.stack_height_change == 1)
+        return InstructionKind::producer;
+    else
+        return InstructionKind::unknown;
+}
+
+struct CodeParams
+{
+    evmc_opcode opcode;
+    Mode mode;
+};
+
+inline constexpr bool operator<(const CodeParams& a, const CodeParams& b) noexcept
+{
+    return std::tuple(a.opcode, a.mode) < std::tuple(b.opcode, b.mode);
+}
+
+std::string to_string(const CodeParams& params)
+{
+    return std::string{instr::traits[params.opcode].name} + '/' +
+           static_cast<char>(get_instruction_kind(params.opcode)) +
+           std::to_string(static_cast<int>(params.mode));
+}
+
+/// Generates the EVM benchmarking loop inner code for the given opcode and "mode".
+bytecode generate_loop_inner_code(CodeParams params)
+{
+    const auto [opcode, mode] = params;
+    const auto kind = get_instruction_kind(opcode);
+    switch (mode)
+    {
+    case Mode::min_stack:
+        switch (kind)
+        {
+        case InstructionKind::push:
+            return stack_limit * (push(opcode, {}) + OP_POP);
+        case InstructionKind::swap:
+        {
+            const auto n = opcode - OP_SWAP1 + 1;
+            return n * OP_DUP1 + stack_limit * 2 * bytecode{opcode} + n * OP_POP;
```

Comment:
If other cases have all `stack_limit` tested opcodes, why does this one have `2 * stack_limit`?

Reply:
To have similar number of instructions in the inner code. This not strictly needed (any number here works), but should allows easier eyeball results comparison (and loop overhead should have similar impact).

I was also considering changing everything to `4*stack_limit` to also use full stack for unops: `DUP1 NOT NOT DUP1 NOT NOT ... POP POP POP ...`.

But it looks like speed difference between "full stack" and "min stack" is due too branch misprediction than to memory cache miss.

Reply:
I too don't really get why some of these have 2*stack_limit.

Reply:
The `2*stack_limit` is your arbitrary number.

Reply:
I see now that it's `stack_limit` for instructions that require POP after each instruction, and `2*stack_limit` for those that don't, so that total count is always `2*stack_limit`.